standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -5390,6 +5406,493 @@ public void countOpenTxns() throws MetaException {
     }
   }
 
+  private void cleanOldStatsFromPartColStatTable(Map<String, PartitionInfo> statsPartInfoMap,
+                                                 Map<String, ColumnStatistics> newStatsMap,
+                                                 Connection dbConn) throws SQLException {
+    PreparedStatement statementDelete = null;
+    int numRows = 0;
+    int maxNumRows = MetastoreConf.getIntVar(conf, ConfVars.DIRECT_SQL_MAX_ELEMENTS_VALUES_CLAUSE);
+    String delete = "DELETE FROM \"PART_COL_STATS\" where \"PART_ID\" = ? AND \"COLUMN_NAME\" = ?";
+
+    try {
+      statementDelete = dbConn.prepareStatement(delete);
+      for (Map.Entry entry : newStatsMap.entrySet()) {
+        // If the partition does not exist (deleted/removed by some other task), no need to update the stats.
+        if (!statsPartInfoMap.containsKey(entry.getKey())) {
+          continue;
+        }
+
+        ColumnStatistics colStats = (ColumnStatistics) entry.getValue();
+        for (ColumnStatisticsObj statisticsObj : colStats.getStatsObj()) {
+          statementDelete.setLong(1, statsPartInfoMap.get(entry.getKey()).partitionId);
+          statementDelete.setString(2, statisticsObj.getColName());
+          numRows++;
+          statementDelete.addBatch();
+          if (numRows == maxNumRows) {
+            statementDelete.executeBatch();
+            numRows = 0;
+            LOG.info("Executed delete " + delete + " for numRows " + numRows);
+          }
+        }
+      }
+
+      if (numRows != 0) {
+        statementDelete.executeBatch();
+      }
+    } finally {
+      closeStmt(statementDelete);
+    }
+  }
+
+  private long getMaxCSId(Connection dbConn) throws SQLException {
+    Statement stmtInt = null;
+    ResultSet rsInt = null;
+    long maxCsId = 0;
+    try {
+      stmtInt = dbConn.createStatement();
+      while (maxCsId == 0) {
+        String query = "SELECT \"NEXT_VAL\" FROM \"SEQUENCE_TABLE\" WHERE \"SEQUENCE_NAME\"= "
```

Comment:
that would create lock on SEQUENCE_TABLE for the duration of the whole stats update operation. Won't it interfere with the regular flow? Could we do this as the last step, and meanwhile do the inserts with UUIDs that would be replaced with  NEXT_VAL? see generateTemporaryId()

Reply:
done